### PR TITLE
Merge price feed wrapper contract

### PR DESCRIPTION
**Pull request by ashneverdawn**
Merging to dia price feed wrapper contract to master as it hadn't been merged yet.
Already includes contract side fixes related to: https://github.com/pendulum-chain/pendulum-ink-wrapper/issues/3

### CODE REVIEW

File: price_feed/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "price_feed"
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
```
Comment (ebma) [inline price_feed/Cargo.toml:4]: Nit
```suggestion
authors = ["Pendulum"]
```

File: price_feed/lib.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    impl PriceFeed {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+        #[ink(message)]
+        pub fn get_coin_info(
+            &mut self,
+            blockchain: [u8; 32],
+            symbol: [u8; 32],
```
Comment (ebma) [inline price_feed/lib.rs:18]: Let's maybe create type aliases for these two and use them also in l. 22 for example.

File: price_feed/lib.rs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        Token(PalletAssetTokenErr),
+        /// An arithmetic error.
+        Arithmetic(PalletAssetArithmeticErr),
+        // unknown error
+        Unknown,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum PalletAssetArithmeticErr {
```
Comment (ebma) [inline price_feed/lib.rs:67]: Nit: lets call it `Error` and not `Err`

File: price_feed/lib.rs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        Overflow,
+        /// Division by zero.
+        DivisionByZero,
+        // unknown error
+        Unknown,
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum PalletAssetTokenErr {
```
Comment (ebma) [inline price_feed/lib.rs:80]: same here

File: price_feed/lib.rs [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        /// There are no providers so the account cannot be created.
+        NoProviders,
+        /// There are too many consumers so the account cannot be created.
+        TooManyConsumers,
+        /// An error to do with tokens.
+        Token(PalletAssetTokenErr),
+        /// An arithmetic error.
+        Arithmetic(PalletAssetArithmeticErr),
+        // unknown error
+        Unknown,
```
Comment (ebma) [inline price_feed/lib.rs:62]: Is this `ChainExtensionError` struct copied/adopted from somewhere or are we introducing it here as we wish? How is `Other` different from `Unknown`? Why would I use one over the other?

File: price_feed/lib.rs [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        Unsupported,
+        // unknown error
+        Unknown,
+    }
+
+    impl ink::env::chain_extension::FromStatusCode for ChainExtensionError {
+        fn from_status_code(status_code: u32) -> Result<(), Self> {
+            match status_code {
+                0 => Ok(()),
+                _ => panic!("encountered unknown status code"),
```
Comment (ebma) [inline price_feed/lib.rs:103]: Should we panic? Is it normal to panic in a smart contract?

File: price_feed/lib.rs [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        use ink_e2e::build_message;
+
+        /// The End-to-End test `Result` type.
+        type E2EResult<T> = std::result::Result<T, Box<dyn std::error::Error>>;
+
+        /// We test that we can upload and instantiate the contract using its default constructor.
+        #[ink_e2e::test]
+        async fn default_works(mut client: ink_e2e::Client<C, E>) -> E2EResult<()> {
+            // Given
+            let constructor = AaaRef::default();
```
Comment (ebma) [inline price_feed/lib.rs:150]: Where is `AaaRef` coming from? Was this taken from an example?

Comment (ebma) [reply price_feed/lib.rs:150]: Can we maybe use something else but 'aaa'? If I understand correctly this is a test instance of our contract?

File: price_feed/lib.rs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        }
+    }
+
+    /// This is how you'd write end-to-end (E2E) or integration tests for ink! contracts.
+    ///
+    /// When running these you need to make sure that you:
+    /// - Compile the tests with the `e2e-tests` feature flag enabled (`--features e2e-tests`)
+    /// - Are running a Substrate node which contains `pallet-contracts` in the background
+    #[cfg(all(test, feature = "e2e-tests"))]
+    mod e2e_tests {
```
Comment (ebma) [inline price_feed/lib.rs:136]: Looks like these are actually related to the 'flipper' sample contract? if so, let's maybe completely remove them for now, doesn't make sense to keep unrelated code, that probably also does not work

**Review by ebma — CHANGES_REQUESTED** [review]

File: price_feed/Cargo.toml (hunk 1 above)
Comment (ashneverdawn) [reply price_feed/Cargo.toml]: done 👍

File: price_feed/lib.rs (hunk 3 above)
Comment (ashneverdawn) [reply price_feed/lib.rs]: Now, referencing the type from pendulum repo directly. It was named `PalletAssetArithmeticError` in there already.

File: price_feed/lib.rs (hunk 6 above)
Comment (ashneverdawn) [reply price_feed/lib.rs]: In a contract, panic will just interrupt that contract's execution. It won't crash the chain or anything. But yes, we should avoid panic in favor of actual errors. Done 👍

File: price_feed/lib.rs (hunk 7 above)
Comment (ashneverdawn) [reply price_feed/lib.rs]: Sorry, this shouldn't have been included. The tests have not yet been implemented. There's another ticket to address this. Removed it from this PR.

File: price_feed/lib.rs (hunk 8 above)
Comment (ashneverdawn) [reply price_feed/lib.rs]: removed 👍

**Review by TorstenStueber — APPROVED** [review]

File: price_feed/lib.rs [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                .input::<(Blockchain, Symbol)>()
+                .output::<Result<CoinInfo, ChainExtError>, false>()
+                .handle_error_code::<ChainExtError>()
+                .call(&input)?
+        }
+    }
+
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct ChainExtError(ChainExtensionError);
```
Comment (ebma) [inline price_feed/lib.rs:32]: The naming here could be a bit confusing but I can't think of something better. Maybe import the `ChainExtensionError` as `RuntimeChainExtensionError` and rename `ChainExtError` to `ChainExtensionError`?
Not that important though

**Review by ebma — APPROVED** [review]
Great, looks way better now 👍
